### ISSUE TITLE
RFC Split out some of .travis.yml to make builds faster

### ISSUE THREAD

**Issue by tractorcow** [silverstripe/silverstripe-framework#8014]
The build matrix for framework is quite complex, and some of the tests being run do bottleneck the overall progress. "time to green" for individual builds can be up to an hour sometimes. For example:

![image](https://user-images.githubusercontent.com/936064/38903358-7ffed608-42f8-11e8-8458-b60e90082b1a.png)

In particular adding coverage adds at least 10 minutes to the overall build time. We don't need coverage on every PR (especially since it's pretty much ignored) but we do use this information.

Another issue is that framework tests running on our travis-ci.org do sometimes block other modules from being built. Especially when we are doing releases, builds can be backed up for quite a long time.

I have looked into possible options:

 - Switching framework entirely to CircleCI. This would prevent other travis / circleci tests from blocking each other, but it is a lot of work to migrate `.travis.yml` to circleci. This also involves choosing, upgrading, and maintaining a docker image, unless we can find a good standard third-party maintained one. Things like behat would need to be re-configured.
 - Switching only low-difficulty tests (non behat, possibly a single matrix with coverage enabled) to circleci. This would probably be a good medium between the above, and doing nothing at all, but we do have the added cost of splitting our maintenance.
 - Remain on travis, but switch sets of tests to daily cron job only. We could filter these jobs (e.g. coverage) to filter based on the `[[ $TRAVIS_EVENT_TYPE = cron ]]` condition. This would mean that certain features (e.g. coverage) wouldn't be available on pull requests anymore, but would be available on request from codecov.
 
I have also investigated paid travis, but it would appear that travis.com is only available for private repos.
 
# PRs:

- [x] https://github.com/silverstripe/silverstripe-framework/pull/8019

**Comment by robbieaverill** [comment]
It's worth mentioning that the coverage run only takes a little longer than the first run in that example, which is a common thing I've seen. Moving coverage runs to a cron would help by removing _some_ of the backlog, but I doubt it will make a very noticeable difference during releases.

I'm against the idea of moving away from Travis for this repo, mainly for the sake of consistency which we're still working to enforce across the SilverStripe supported modules.

We might be able to separate the testsuite up into different sections that can be run in parallel (with or without coverage). I notice the Travis pricing page says "contact us to discuss your custom requirements" - perhaps we could find out about higher parallelism for open source projects. Either way, a higher number of concurrent builds will help to solve these problems without having to worry too much about speeding up the test suite in general (that's another story altogether!)

**Comment by tractorcow** [comment]
The most major parallel issue is running multiple builds on the same repo. If we have 3 PRs on framework only one will run at one time. parallel matrixes within each build itself isn't the main issue.

**Comment by robbieaverill** [comment]
I have a suspicion that’s not quite true, I’ve seen jobs within a build start to kick off as the queue frees up, rather that all of them starting at once for a build. Would be good to see what Travis have to say :-)

**Comment by dhensby** [comment]
> The most major parallel issue is running multiple builds on the same repo. If we have 3 PRs on framework only one will run at one time. parallel matrixes within each build itself isn't the main issue.

Yeah, this isn't true. They are shared across an organisation (so SilverStripe/*) and we can use as many workers as we have either on the same repo or different ones in our org.

**Comment by dhensby** [comment]
1. Can we use another service for coverage itself to free up travis time?
2. Can we only run coverage *if* the others pass (how much time is wasted needlessly generating coverage reports?)
3. Do we even need coverage? As you say it's mostly ignored on PRs, if we aren't looking at it or blocking merges on it, then what's the use? I've found coverage to be a bit misleading too as I've seen PRs remove code and coverage has gone *down*!
4. How about moving CWP modules out of the SS org so we reduce the contention ratio on the silverstripe travis account?

**Comment by tractorcow** [comment]
PR to make coverage conditional at https://github.com/silverstripe/silverstripe-framework/pull/8019

**Comment by maxime-rainville** [comment]
I've merged Damien's PR #8019, but looks like you guys are still discussing this issue. I'll leave this one open for now.

**Comment by chillu** [comment]
Damian recommended to close this, given the PR is merged. There's more we could do, if there's specific actions feel free to reopen.
